bundler/spec/commands/fund_spec.rb

```diff
@@ -55,6 +55,45 @@
     expect(out).to_not include("gem_with_dependent_funding")
   end
 
+  it "does not consider fund information for uninstalled optional dependencies" do
+    install_gemfile <<-G
+      source "#{file_uri_for(gem_repo2)}"
+      group :whatever, optional: true do
+        gem 'has_funding_and_other_metadata'
+      end
+      gem 'has_funding'
+      gem 'rack-obama'
+    G
+
+    bundle "fund"
+
+    expect(out).to include("* has_funding (1.2.3)\n  Funding: https://example.com/has_funding/funding")
+    expect(out).to_not include("has_funding_and_other_metadata")
+    expect(out).to_not include("rack-obama")
+  end
+
+  it "considers fund information for installed optional dependencies" do
+    install_gemfile <<-G
+      source "#{file_uri_for(gem_repo2)}"
+      group :whatever, optional: true do
+        gem 'has_funding_and_other_metadata'
+      end
+      gem 'has_funding'
+      gem 'rack-obama'
+    G
+
+    ENV["BUNDLE_WITH"] = "whatever"
```

Comment:
You can instead run `bundle config with whatever` before `install_gemfile` above, so you don't need to `bundle install` again nor to clean up the `ENV` variable in an `ensure` block (I think it's not necessary anyways since the global `after` hook will reset `ENV` anyways).